File: package/java/src/main/java/com/programknock/ServerLogAnalysis.java
package com.programknock;

import java.util.*;
import java.util.stream.Collectors;

public class ServerLogAnalysis {

    public static class LogEntry {
        public final String ipAddress;
        public final String request;
        public final int statusCode;

        public LogEntry(String ipAddress, String request, int statusCode) {
            this.ipAddress = ipAddress;
            this.request = request;
            this.statusCode = statusCode;
        }

        @Override
        public boolean equals(Object obj) {
            if (this == obj) return true;
            if (obj == null || getClass() != obj.getClass()) return false;
            LogEntry logEntry = (LogEntry) obj;
            return statusCode == logEntry.statusCode &&
                   Objects.equals(ipAddress, logEntry.ipAddress) &&
                   Objects.equals(request, logEntry.request);
        }

        @Override
        public int hashCode() {
            return Objects.hash(ipAddress, request, statusCode);
        }
    }

    public static List<LogEntry> filterSuccessfulRequests(List<LogEntry> logs) {
        return logs.stream()
            .filter(log -> log.statusCode == 200)
            .collect(Collectors.toList());
    }

    public static Map<String, Integer> countRequestsByIp(List<LogEntry> logs) {
        Map<String, Integer> ipCounts = new HashMap<>();
        for (LogEntry log : logs) {
            ipCounts.merge(log.ipAddress, 1, Integer::sum);
        }
        return ipCounts;
    }
}

File: package/php/src/ServerLogAnalysis.php
<?php

declare(strict_types=1);

namespace ProgramKnock;

class ServerLogAnalysis
{
    /**
     * ステータスコードが200のリクエストをすべて抽出
     *
     * @param array<array{0: string, 1: string, 2: int}> $logs ログデータのリスト
     * @return array<array{0: string, 1: string, 2: int}> ステータスコードが200のログのリスト
     */
    public static function filterSuccessfulRequests(array $logs): array
    {
        $successfulRequests = [];

        foreach ($logs as $log) {
            if ($log[2] === 200) {
                $successfulRequests[] = $log;
            }
        }

        return $successfulRequests;
    }

    /**
     * 各IPアドレスのリクエスト回数を集計
     *
     * @param array<array{0: string, 1: string, 2: int}> $logs ログデータのリスト
     * @return array<string, int> IPアドレスごとのリクエスト回数の辞書
     */
    public static function countRequestsByIp(array $logs): array
    {
        $ipCounts = [];

        foreach ($logs as $log) {
            $ip = $log[0];

            if (isset($ipCounts[$ip])) {
                $ipCounts[$ip]++;
            } else {
                $ipCounts[$ip] = 1;
            }
        }

        return $ipCounts;
    }
}
